src/Engine/SpriteAnimator/SpriteAnimator.cpp

```diff
@@ -0,0 +1,70 @@
+#include "SpriteAnimator.h"
+
+SpriteAnimator::SpriteAnimator()
+    : initialized_(false)
+{}
+
+// Keeping this as an init right now because the objects may not
+int SpriteAnimator::init(SDL_Texture* texture, int rows, int cols, int on_every) {
+    //If there was no draw_order given as an argument, we will create a default
+    if (draw_order_.size() == 0) {
+        for (int i = 0; i < rows*cols; i++) {
+            draw_order_.push_back(i);
+        }
+    }
+    on_every_ = on_every;
+    rows_ = rows;
+    cols_ = cols;
+    //total_frames_ = rows * cols;
+    current_index_ = 0;
+    int w, h;
+    SDL_QueryTexture(texture, NULL, NULL, &w, &h);
+    // We will run off the assumption that the given
+    // spritesheet has it's starting sprites on 0,0.
+    // If there is padding... manually get rid of it for now
+    total_sprite_sheet_size_.x = 0;
+    total_sprite_sheet_size_.y = 0;
+    total_sprite_sheet_size_.w = w;
+    total_sprite_sheet_size_.h = h;
+    initialized_ = true;
+    return 0;
+}
+
+int SpriteAnimator::init(SDL_Texture* texture, int rows, int cols, int on_every, std::vector<int> draw_order) {
+    draw_order_ = draw_order;
+    return init(texture, rows, cols, on_every);
+}
+
+
+SDL_Rect SpriteAnimator::get_next_frame(int& returned_frame) {
+    // We want to be able to call this function, and start with frame 0
+    // So we advance the frame at the end of this function
+    returned_frame = draw_order_.at(current_index_);
+    SDL_Rect return_rect;
+
+    int on_row = draw_order_.at(current_index_) / cols_; //integer division for the floor
+    int on_col = draw_order_.at(current_index_) % (rows_ - 1);
+
+    return_rect.w = total_sprite_sheet_size_.w / cols_;
+    return_rect.h = total_sprite_sheet_size_.h / rows_;
+    return_rect.x = on_col * (total_sprite_sheet_size_.w / cols_);
+    return_rect.y = on_row * (total_sprite_sheet_size_.h / rows_);
```

Comment:
What do you think of creating a vector that precalculates all this in the init, a map of a rect to a frame? So then grabbing next frame is alot simpler.

Reply:
I think it's ok, but it's not really an optimization, and we shouldn't optimize lots of these things for now.

The reason why it's not an optimization (even though it'll be in init), is because init happens during the gameplay time, and not during the "init" section of the states (which we don't have, other than commenting).

If the player shoots an enemy, they create a `missile`, the missile then goes through it's `init`, where it'll have to pre-calculate for ALL of it's things, right away, rather than doing a single calc every loop. So this, probably won't be considered a optimization, and if we want to optimize it, this extra work that I'll do, will be erased anyways.

So i'm opting not to do it for now because it won't be an optimization.